### PR TITLE
💻 Computed Fields

### PR DESCRIPTION
**Pull request by awtkns**
Supports NOOP, COALESCE, UPPER, LOWER, and SLUGIFY. Allows for chaining and complex nested actions. eg:

```
LOWER(CONCAT(UPPER('hello'), ' ', CONCAT(LOWER(UPPER(name))), ' from ', city.name))
```

Inspired from the SQL dsl. We need a dsl as we dont want to exec arbitrary code here. It is also much faster then using eval which spins up interpreter.

### CODE REVIEW

File: harambe/parser/expression/functions.py
Comment (awtkns) [inline harambe/parser/expression/functions.py:1]: Very easy to add functions in the future

File: harambe/parser/expression/evaluator.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    def evaluate(expression: str, obj: Any) -> Any:
+        expression = expression.strip()
+
+        if not expression:
+            raise ValueError("Empty expression")
+
+        if "(" not in expression and ")" not in expression:
+            return ExpressionEvaluator._get_field_value(expression, obj)
+
+        func_name = expression.split("(", 1)[0].strip().upper()
```
Comment (asim-shrestha) [inline harambe/parser/expression/evaluator.py:30]: image

will need to limit key characters first lolol. The above is a valid schema key right now which will fuk this all up

File: test/parser/test_parser.py [hunk 2]
```diff
@@ -583,3 +583,30 @@ def test_dump_email() -> None:
...
+
+
+def test_computed_field() -> None:
+    schema = {
+        "degree": {"type": "string"},
+        "first_name": {"type": "string"},
+        "last_name": {"type": "string"},
+        "full_name": {
+            "type": "string",
+            "expression": "CONCAT(first_name, ' ', last_name)",
```
Comment (asim-shrestha) [inline test/parser/test_parser.py:595]: we'd also need to parse this stuff in the front end which might be annoying. Should we make this an object?

Comment (asim-shrestha) [reply test/parser/test_parser.py:595]:
```suggestion
            "expression": {
                "name": "CONCAT"
                "params": [
                        {"name": "CONCAT", "params": ["'test'", " ", "'zest'"]},
                        "field.other_field"
                .........
```

File: harambe/parser/expression/evaluator.py (hunk 1 above)
Comment (awtkns) [reply harambe/parser/expression/evaluator.py:30]: yeah will do that as a follow up.

File: test/parser/test_parser.py (hunk 2 above)
Comment (awtkns) [reply test/parser/test_parser.py:595]: Decided to leave as string so the users can "program"

**Pull request merged**